Dummy template RBT on <T>

amcastro-tri opened the pull request:
This performs the first step of the templatization of `RBT` on `<T>` as outlined in #3987.

This PR does not change or add any functionality. It is only a refactor from `RBT` to `RBT<T>` and it should be simple to review even though it touches about 140 files.

 After this PR `RigidBodyTree` instantiations will need to be spelled out with the particular type i.e. `RigidBodyTree<double>`, `RigidBodyTree<AutoDiffScalar<Vector3d>>` for instance. 

All System 2.0 components already templated on `<T>` also template RigidBodyTree on `<T>`. For instance `RigidBodyPlant`. 
All other code that right now uses RBT methods with `<double>` are updated to simply instantiate a `RigidBodyTree<double>`.

The only "actual" difference (beside changing `RBT` by `RBT<T>`) is outlined below:
I had to split the implementations in `ConstraintWrappers.h` (now `constraint_wrappers.h`) into their own source in `constraint_wrappers.cc` so that they'd get compiled into objects and I wouldn't get undefined references at link time.
This required to place a `DRAKE_EXPORT` in the header `constraint_wrappers.h` instead of at the place of explicit instantiation in `constraint_wrappers.cc` to make the `KinematicsCacheHelper<T>`'s constructor visible to `drakeRBSystem` or otherwise I would get a link error.
Notice also this required to make explicit the dependency of `drakeRBSystem` on `drakeIK` in `drake/systems/plants/CMakeLists.txt`.

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3940)

<!-- Reviewable:end -->

amcastro-tri commented:
Reviewed 13 of 14 files at r1.
Review status: 13 of 14 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-debug please
@drake-jenkins-bot linux-clang-ninja-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please
@drake-jenkins-bot linux-gcc-ninja-experimental-open-source-debug please

amcastro-tri commented:
Review status: 3 of 138 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWPf6BeQKofqa6LyM2:-KVWPf6BeQKofqa6LyM3:boia80t) (no related file):*
+@david-german-tri for platform review please
+@sherm1 for feature review please.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 3 of 139 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWPf6BeQKofqa6LyM2:-KVWn0y15_uOu5FXyHii:b-cc8g0m) (no related file):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> +@david-german-tri for platform review please
> +@sherm1 for feature review please.
</details>
@amcastro-tri, this PR has conflicts with master that need to be merged manually.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint. There are a few places where `RBTree<double>` is adjacent to `KinematicsCache<Scalar>`. I *think* I understand why that is but BTW'ed some of them to make sure that's right.

---

Reviewed 42 of 134 files at r2, 11 of 52 files at r3.
Review status: 56 of 139 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/QPInverseDynamicsForHumanoids/rigid_body_tree_utils.cc, line 10 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWseoT2x56CHRA0Is2:-KVWseoT2x56CHRA0Is3:bdu5ho8) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/examples/QPInverseDynamicsForHumanoids/rigid_body_tree_utils.cc#L10)):*
> ```cpp
> 
> Eigen::Vector6d GetTaskSpaceVel(const RigidBodyTree<double>& r,
>                                 const KinematicsCache<double>& cache,
> ```

BTW, I like how `RigidBodyTree<double>` now matches `KinematicsCache<double>` here.

---

*[drake/matlab/systems/plants/contactConstraintsmex.cpp, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWtPogPpSakEcE5hto:-KVWtPogPpSakEcE5htp:beaswxb) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/contactConstraintsmex.cpp#L43)):*
> ```cpp
> template <typename Scalar>
> pair<MatrixX<Scalar>, vector<MatrixX<Scalar>>> contactConstraints(
>     const RigidBodyTree<double> &model, const KinematicsCache<Scalar> &cache,
> ```

BTW, should this be Scalar here?

---

*[drake/matlab/systems/plants/createKinematicsCacheAutoDiffmex.cpp, line 8 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWtdYbJhX5zFRPegaI:-KVWtdYbJhX5zFRPegaJ:beaswxb) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/createKinematicsCacheAutoDiffmex.cpp#L8)):*
> ```cpp
> 
> template <typename Scalar>
> mxArray *createKinematicsCache(RigidBodyTree<double> &model) {
> ```

BTW, should this be Scalar here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVWtx-NR3mBQmo9JvtB:b-miorlk)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint2. I am confused about the use of `RigidBodyTreed` vs. `RigidBodyTree<double>`. I remember you were going to define `RigidBodyTreed` for Director to use but I'm seeing it in a subset of the files here, particular mex files and parser-related files. Why? Also a few comments sprinkled below, PTAL.

---

Reviewed 1 of 14 files at r1, 23 of 134 files at r2, 28 of 52 files at r3.
Review status: 107 of 139 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWuXyqPn6laoVS7IPs:-KVWuXyqPn6laoVS7IPt:bdunzyn) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp#L27)):*
> ```cpp
>   auto func_double =
>       make_function(
>           &RigidBodyTreed::centerOfMassJacobianDotTimesV<double>);
> ```

BTW, why `RigidBodyTreed` here but `RigidBodyTree<double>` elsewhere?

---

*[drake/systems/plants/constraint_wrappers.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWwah58CFDQY91by05:-KVWwah67y4l8uGhKlSm:bt8kvy1) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/constraint_wrappers.h#L26)):*
> ```objc
> /// TODO(amcastro-tri): DRAKE_EXPORT is used here in the header instead of at
> /// the place of explicit instantiation. Explore a solution for this problem as
> /// described in #3940.
> ```

BTW, there should be no need for DRAKE_EXPORT at all after #3965 re-lands (it was merged but is pending reversion in #3981).

---

*[drake/systems/plants/parser_common.cc, line 24 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWxZr9ZigpsKNS6Dy8:-KVWxZrAq0xazwzSbgb_:bjhgfo5) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/parser_common.cc#L24)):*
> ```cpp
>     const std::shared_ptr<RigidBodyFrame> weld_to_frame,
>     const PoseMap* pose_map,
>     RigidBodyTreed* tree) {
> ```

BTW, why `RigidBodyTreed` here? (And in related files.)

---

*[drake/systems/plants/RigidBodyTree.cpp, line 2089 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWymT5Jg1XqqsfUbyb:-KVWymT5Jg1XqqsfUbyc:b-yrgwvz) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.cpp#L2089)):*
> ```cpp
>       } else {
>         throw std::logic_error(
>             "RigidBodyTree<T>::FindBody: ERROR: found multiple bodys named \"" +
> ```

I don't think there should be a `<T>` in the error message. You could replace it with the name of the actual type but I don't think that's worth the trouble; just remove it. (Here and elsewhere in this file.)

---

*[drake/systems/plants/RigidBodyTree.h, line 403 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWzW7wmqm6YtTRdVDP:-KVWzW7wmqm6YtTRdVDQ:b7rrlo5) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.h#L403)):*
> ```objc
>   /// inverseDynamics and dynamicsBiasTerm.
>   using BodyToWrenchMap = drake::eigen_aligned_std_unordered_map<
>     RigidBody const*, drake::WrenchVector<T>>;
> ```

BTW, I see you converted this to T but I thought you weren't going to do that until a later PR. Is this alias intentionally treated differently than the templatized methods?

---

*[drake/systems/plants/RigidBodyTree.h, line 476 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWzy25bh13vCYj9nJv:-KVWzy25bh13vCYj9nJw:b2axn1z) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.h#L476)):*
> ```objc
>                       DerivedPoints::RowsAtCompileTime == Eigen::Dynamic,
>                   "points argument has wrong number of rows");
>     auto Tr =
> ```

BTW, a better name would be nice.

---

*[drake/systems/plants/RigidBodyTreeContact.cpp, line 306 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX-RPbvhS9BbjpuGbk:-KVX-RPbvhS9BbjpuGbl:b-25oswi) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTreeContact.cpp#L306)):*
> ```cpp
>     Eigen::Matrix<double, -1, -1, 0, -1, -1> &) const;
> 
> template DRAKE_EXPORT void RigidBodyTree<double>::surfaceTangents(
> ```

BTW, what is this doing here? Looks like a forward declaration?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVX01STK-9yjZLu29wz:blfthio)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done. Wow -- you deserve a combat medal for this, Alejandro! Or maybe some company-paid counseling to help you recover.

---

Reviewed 2 of 14 files at r1, 19 of 134 files at r2, 13 of 52 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/test/benchmarkRigidBodyTree.cpp, line 83 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX13lBNh7tShNgrVjn:-KVX13lBNh7tShNgrVjo:bbt8wol) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/benchmarkRigidBodyTree.cpp#L83)):*
> ```cpp
>   uniform_real_distribution<> uniform(0, 1);
> 
>   const typename RigidBodyTree<Scalar>::BodyToWrenchMap no_external_wrenches;
> ```

BTW, is this correct?

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1SqlMD9rGjE5PKXD:-KVX1SqlMD9rGjE5PKXE:b1l0ddw) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/urdf_parser_test.cc#L53)):*
> ```cpp
>   // Instantiates a RigidBodyTree using the URDF string defined above.
>   std::unique_ptr<RigidBodyTree<double>>
>       rigid_body_tree(new RigidBodyTree<double>());
> ```

BTW, other parsing code used `RigidBodyTreed` -- I'm not sure why, and not sure why it isn't the same here.

---

*[drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1miWxeY2JiGYV58V:-KVX1miWxeY2JiGYV58W:brjhw9n) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc#L118)):*
> ```cpp
>     tree_rpy_->doKinematics(kinematics_cache_coriolis, true);
> 
>     const typename RigidBodyTree<Scalar>::BodyToWrenchMap no_external_wrenches;
> ```

BTW, this is OK?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVX29W2mvDueEINUr-z:bz443mp)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you for the review @sherm1. I think you deserve a medal for reviewing it! You raised some very good points that I tried address (mostly with answers only at this point, no code, sorry). However I made sure to add them to my list of pain points for the next check-mark in #3987.

You essentially raised two major points: 
1. Why am I using `RigidBodyTree<double>` in code using auto diff methods (like in mex funcions)?. For mex functions the reason is that the Matlab supported RBT version is consistent with what we call now `RBT<double>` i.e. no differentiation with respect to parameters possible but just state variables. The only way to deal with that is making actual changes to mex files to suport that capability which of course is out of the scope of this PR (similarly with updates to `rbtree.i` supporting the python API).
2. Why `RigidBodyTreed` in parser and mex files and `RigidBodyTree<double>` elsewhere? no good answer but the fact that using `RigidBodyTree<double>` created a lot of lint errors for these files. When prototyping this that was the fastest approach for me to move forward. Now we know it can compile and run I could change those to also use `RigidBodyTree<double>` if you feel strong about that.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/matlab/systems/plants/contactConstraintsmex.cpp, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWtPogPpSakEcE5hto:-KV_B_l3RpxlYCoCWW8L:b4clzuv) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/contactConstraintsmex.cpp#L43)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, should this be Scalar here?
</details>
The basic idea is that I kept old code that does not differentiate with respect to parameters to use `<double>`. I do agree that in this case however using `<Scalar>` makes more sense. In particular, when tackling item two of the list in #3987 I will need to do exactly what you suggest.
However, when trying that now I get a horrible chain reaction of errors at compile time. The problem seems to be that `RigidBodyTree<AutoDiffScalar<....>>` is "not mex convertible".
I would need to investigate that further and I think this PR is big enough already and I'm afraid delaying its merge into master will make it difficult to keep in sync. 
I added a link to this comment in #3987.

---

*[drake/matlab/systems/plants/createKinematicsCacheAutoDiffmex.cpp, line 8 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWtdYbJhX5zFRPegaI:-KV_DlnjqaSYmvVkrVp4:b1xyvsf) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/createKinematicsCacheAutoDiffmex.cpp#L8)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, should this be Scalar here?
</details>
See usage of this method below in `mexFunction` in this file. Notice that in line 48 there is a call to `getDrakeMexPointer`. This method will return a pointer to a version of `RBT` that is consistent to what in this PR we call `RBT<double>`. Trying to change that in this PR goes more along the lines of "implement  Matlab bindings supporting auto differentation with respect to parameters", which of course it out of the scope of this PR. 

---

*[drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWuXyqPn6laoVS7IPs:-KV_FvYQQOBxwKq1MptB:b-60rg4q) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, why `RigidBodyTreed` here but `RigidBodyTree<double>` elsewhere?
</details>
There are tons of mex functions and, as I did with parsers, I decided to use the `RigidBodyTreed` typedef to avoid having to fix a huge number of lint errors.
I can change to `RigidBodyTree<double>` if you feel strong about it.

---

*[drake/systems/plants/constraint_wrappers.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWwah58CFDQY91by05:-KV_GdoNTOD_X1_WrFFe:b-uvfws6) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/constraint_wrappers.h#L26)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, there should be no need for DRAKE_EXPORT at all after #3965 re-lands (it was merged but is pending reversion in #3981).
</details>
I will wait for that to land then. In the meantime this is still valid.

---

*[drake/systems/plants/parser_common.cc, line 24 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWxZr9ZigpsKNS6Dy8:-KV_GpHn-XDsWFPMi2CJ:b-xgru71) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/parser_common.cc#L24)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, why `RigidBodyTreed` here? (And in related files.)
</details>
As I mentioned for mex files. These are such a large list of methods with usually many arguments that when changing to `RigidBodyTree<double>` cause a very large amount of lint errors (the typical <=80 one).
I can change to `RigidBodyTree<double>` if you feel strong about it. 

---

*[drake/systems/plants/RigidBodyTree.cpp, line 2089 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWymT5Jg1XqqsfUbyb:-KV_HpOH2WEby1HkNkqQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.cpp#L2089)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I don't think there should be a `<T>` in the error message. You could replace it with the name of the actual type but I don't think that's worth the trouble; just remove it. (Here and elsewhere in this file.)
</details>
Done.

---

*[drake/systems/plants/RigidBodyTree.h, line 403 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWzW7wmqm6YtTRdVDP:-KV_I1Psh2GDzLJAMYxa:b-yes68e) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.h#L403)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, I see you converted this to T but I thought you weren't going to do that until a later PR. Is this alias intentionally treated differently than the templatized methods?
</details>
I remember this particular piece of code causing a lot of trouble when having the double template on `<T>` and on `<Scalar>` though to be honest I don't remember the details (maybe the explicit instantiation but again, I'm fuzzy after modifying over 100 files).
However, this solution is the one solution we want in the code base and, it reads way more nicely. So I decided to go with this.

---

*[drake/systems/plants/RigidBodyTree.h, line 476 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWzy25bh13vCYj9nJv:-KV_JZX_YMEo9dmYYYrC:bw2ba11) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.h#L476)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, a better name would be nice.
</details>
I changed the name and added some comments to clarify. It is difficult to come up with a good short name when your arguments are called things like "from_body_or_frame_ind".

---

*[drake/systems/plants/RigidBodyTreeContact.cpp, line 306 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX-RPbvhS9BbjpuGbk:-KV_K3itYYHEojC1enT9:b-mokft1) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTreeContact.cpp#L306)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, what is this doing here? Looks like a forward declaration?
</details>
BTW. It is an explicit instantiation as the ones right above. It was not needed when RBT had no template and the explicit instantiation of RBT itself happens in another file (RigidBodyTree.cpp). This will disappear  when I change `<Scalar>` to `<T>` and everything gets explicitly instantiated in one single place when explicitly instantiating RBT itself.

---

*[drake/systems/plants/test/benchmarkRigidBodyTree.cpp, line 83 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX13lBNh7tShNgrVjn:-KV_L1KQrk63l-qBVQI_:bhlqcpm) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/benchmarkRigidBodyTree.cpp#L83)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, is this correct?
</details>
Yes, it is. I could try to go one step further and modify this method to use `RigidBodyTree<Scalar>` but I am very sure that will fail somewhere. That again goes along the idea of updating old code to use and API that would allow to auto diff in parameters.
Here, and in the places you identified in the mex files, I think we first need some version of `cast<Scalar>` or `clone<Scalar>` before in order to make the appropriate conversions. 
Doing that in this PR I think would cause a lot of extra pain.
Added this to the list of pain points together with mex files in #3987. 

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1SqlMD9rGjE5PKXD:-KV_Mj1eoU5XwcN0nMNS:bkqyxpa) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/urdf_parser_test.cc#L53)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, other parsing code used `RigidBodyTreed` -- I'm not sure why, and not sure why it isn't the same here.
</details>
Ne reason really more than the parser files generating huge amounts of lint when using `RigidBodyTree<double>`. 

---

*[drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1miWxeY2JiGYV58V:-KV_NEq6j3eeJfcxozFD:b-z6zklm) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc#L118)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, this is OK?
</details>
Yes. I am again using this "trick" of having a `RigidBodyTree<double>` whenever we are not differentiating with respect to parameters (which of course we never do). Otherwise I would need to actually make modifications in this test to have two instances of the same tree (double and autodiff). 
This piece of code will be one of the first to revisit when templating methods. I think we will need the `cast<Scalar>` method for this to read nicely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KV_Nn1y7eVMEN2mQ4SW:b-hsu31v)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
I didn't look at all of the boring files; I'll trust you + CI on that.  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, and kudos.

+(status: curate commits before merging) 

---

Reviewed 1 of 14 files at r1, 11 of 134 files at r2, 10 of 52 files at r3, 4 of 6 files at r4.
Review status: 137 of 139 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.h, line 158 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVacI2ayAF9j5QvGTl0:-KVacI2b6vdRNmqLBrFp:b-iidlyp) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/RigidBodySystem.h#L158)):*
> ```objc
> 
>   explicit RigidBodySystem(
>       std::shared_ptr<RigidBodyTree<double>> rigid_body_tree)
> ```

BTW to check my understanding, will this be RigidBodyTree<T> in future work?

---

*[drake/systems/plants/RigidBodyTree.cpp, line 859 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVabrB98HnSeDlLrJQN:-KVabrBAZ-k9cDSq38JX:b-e3i5mn) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/RigidBodyTree.cpp#L859)):*
> ```cpp
> 
> template <typename T>
> template <typename Scalar>
> ```

Are these actually supposed to be two different types, or is that just a temporary state?  If the latter, did you consider `RigidBodyTree<Scalar>`?

META for several other lines below.

---

*[drake/systems/plants/constraint/RigidBodyConstraint.cpp, line 999 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVad-vwUN-yx4OwUQkc:-KVad-vwUN-yx4OwUQkd:bu9kmn5) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/constraint/RigidBodyConstraint.cpp#L999)):*
> ```cpp
> }
> 
> WorldEulerConstraint::WorldEulerConstraint(RigidBodyTree<double>* robot, int body,
> ```

Several >80 in this file?

---

*[drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1miWxeY2JiGYV58V:-KVaeG2uRKNtGTk9u8XA:b-uv7wjw) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc#L118)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes. I am again using this "trick" of having a `RigidBodyTree<double>` whenever we are not differentiating with respect to parameters (which of course we never do). Otherwise I would need to actually make modifications in this test to have two instances of the same tree (double and autodiff). 
> This piece of code will be one of the first to revisit when templating methods. I think we will need the `cast<Scalar>` method for this to read nicely.
</details>
FYI, a sticking point with `cast<Scalar>` is that some of the possible combinations are meaningless (how do you cast a `SymbolicExpression` to `double`?)  The approach we settled on for `System<T>` in #3967 is explicit conversion functions like `ToAutoDiffXd(System<double>&)`.  Might be useful for you here as well.  (Totally agree you shouldn't mess with it in this PR, regardless.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVag4E_pXtbIQfNONVZ:bd19stt)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 137 of 139 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.h, line 158 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVacI2ayAF9j5QvGTl0:-KVaiOdO9DSSB7PuJEUH:b-bjyy9) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/RigidBodySystem.h#L158)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW to check my understanding, will this be RigidBodyTree<T> in future work?
</details>
BTW, my understanding is that we will get rid of all system 1.0 code. 
For RigidBodyPlant<T> (and any other <T> classes) the plan is to use `RigidBodyTree<T>`. Two examples already in this PR are `RigidBodyPlant<T>` and `AutomotiveSimulator<T>`.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 859 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVabrB98HnSeDlLrJQN:-KVaip7gSYGhN5eMoyZs:b-r34kem) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/RigidBodyTree.cpp#L859)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Are these actually supposed to be two different types, or is that just a temporary state?  If the latter, did you consider `RigidBodyTree<Scalar>`?
> 
> META for several other lines below.
</details>
Please look at #3987 for my planned work. In a nutshell, this is a "very simple" PR that templates RBT on a "dummy" T, meaning that I am still leaving all the temlatized methods in RBT still templated on the arguments `<Scalar>` type. 
The next step will be to get rid of `<Scalar>` all over the place, template methods on `<T>` and get rid of each explicit instantiation for each method since a single instantiation on `RBT<T>` will suffice.

---

*[drake/systems/plants/constraint/RigidBodyConstraint.cpp, line 999 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVad-vwUN-yx4OwUQkc:-KVakHCpRfX3GSBVbqJF:b-sb7ev2) ([raw file](https://github.com/robotlocomotion/drake/blob/32c34f3d8231388984e27660f73a92416b979949/drake/systems/plants/constraint/RigidBodyConstraint.cpp#L999)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Several >80 in this file?
</details>
Ah, I didn't realize this file was excluded from CPPLINT.cfg. Fixed now.

---

*[drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVX1miWxeY2JiGYV58V:-KVakbOtapEZCoMKZrKd:bhvth95) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/test/rigid_body_tree/rigid_body_tree_inverse_dynamics_test.cc#L118)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> FYI, a sticking point with `cast<Scalar>` is that some of the possible combinations are meaningless (how do you cast a `SymbolicExpression` to `double`?)  The approach we settled on for `System<T>` in #3967 is explicit conversion functions like `ToAutoDiffXd(System<double>&)`.  Might be useful for you here as well.  (Totally agree you shouldn't mess with it in this PR, regardless.)
</details>
Thank you for the link to your PR!. I will for sure take a look.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 14 files at r1, 72 of 134 files at r2, 37 of 52 files at r3, 2 of 6 files at r4.
Review status: 138 of 139 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVanUMUBxcTrbE1XQBR:bg7igkv)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done. Just a couple of comments left, PTAL.

---

Reviewed 6 of 6 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWuXyqPn6laoVS7IPs:-KVb6R4Vdg6SMNJB4Hj2:b-xtuan) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp#L27)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> There are tons of mex functions and, as I did with parsers, I decided to use the `RigidBodyTreed` typedef to avoid having to fix a huge number of lint errors.
> I can change to `RigidBodyTree<double>` if you feel strong about it.
</details>
I'm happy with this as long as `RigidBodyTreed` is some local typedef abbreviation for `RigidBodyTree<double>`. If it is a global then I think it should be used consistently everywhere.

---

*[drake/systems/plants/parser_common.cc, line 24 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWxZr9ZigpsKNS6Dy8:-KVb4PnKJvOX8xrZUcWY:b-ukzuzw) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/parser_common.cc#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> As I mentioned for mex files. These are such a large list of methods with usually many arguments that when changing to `RigidBodyTree<double>` cause a very large amount of lint errors (the typical <=80 one).
> I can change to `RigidBodyTree<double>` if you feel strong about it. 
</details>
Oh, I see. Is `RigidBodyTreed` just a local typedef then in this file? That would be OK I think. I was worried that this is the same global `RigidBodyTreed` that you added for Director. If it's referring to something global I think you should fix it, otherwise it's fine.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 2089 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWymT5Jg1XqqsfUbyb:-KVb4nTgVgBjQABBNzgT:b5pa0hh) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.cpp#L2089)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done.
</details>
no change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVb73Xr39ekCvmv8O6v:b-8umgrw)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@sherm1, I addressed your comments. PTAL

---

Review status: 126 of 139 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWuXyqPn6laoVS7IPs:-KVeT96PRXtEufUBR5OD:btoznyj) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/matlab/systems/plants/rigidBodyTreeMexFunctions.cpp#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I'm happy with this as long as `RigidBodyTreed` is some local typedef abbreviation for `RigidBodyTree<double>`. If it is a global then I think it should be used consistently everywhere.
</details>
Done. Now `RigidBodyTreed` is only local to super lengthy mex files (rigidBodyTreeMexFunctions.cpp, constructPtrRigidBodyConstraintsmex.cpp and updatePtrRigidBodyConstraintmex.cpp)

---

*[drake/systems/plants/parser_common.cc, line 24 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWxZr9ZigpsKNS6Dy8:-KVeTn8HX4FkoE0zYxmT:bmg9764) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/parser_common.cc#L24)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Oh, I see. Is `RigidBodyTreed` just a local typedef then in this file? That would be OK I think. I was worried that this is the same global `RigidBodyTreed` that you added for Director. If it's referring to something global I think you should fix it, otherwise it's fine.
</details>
No, it is not local, it is defined in `RigidBodyTree.h` and yes, it is used by director. This is similar to having `VectorXd`. Why does it bother you to have it "globaly" in `RigidBodyTree.h`?.
In any case. I did modify this file to use `RigidBodyTree<double>`. Now only mex files use RigidBodyTreed.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 2089 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWymT5Jg1XqqsfUbyb:-KVeV1rUL2Llk247unYE:b5qnaz3) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/RigidBodyTree.cpp#L2089)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> no change?
</details>
Sorry, I only made the change inside comments. I included string literals now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVeVOOoWqoebv-4LjRj:bv009m1)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/parser_common.cc, line 24 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-KVWxZr9ZigpsKNS6Dy8:-KVeoeQEiFsjQ9LYpNDr:b-xlpf8p) ([raw file](https://github.com/robotlocomotion/drake/blob/ad9ff116dbdc3a6a78dfa6d87487b9ca4429f486/drake/systems/plants/parser_common.cc#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> No, it is not local, it is defined in `RigidBodyTree.h` and yes, it is used by director. This is similar to having `VectorXd`. Why does it bother you to have it "globaly" in `RigidBodyTree.h`?.
> In any case. I did modify this file to use `RigidBodyTree<double>`. Now only mex files use RigidBodyTreed.
</details>
Thanks. There were a few things that were bothering me:
- I didn't understand if this actually had the same meaning as the declaration you made for Director.
- It wouldn't be easy to convert this later to `RB<T>`.
- It wasn't used consistently throughout the code.

Thanks for changing it -- I feel better now!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3940#-:-KVeppwqRcLVk2XzAYfB:bnfp4nl)*
<!-- Sent from Reviewable.io -->
